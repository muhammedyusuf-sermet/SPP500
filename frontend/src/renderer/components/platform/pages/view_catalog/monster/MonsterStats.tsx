import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';
import { Tile, Subtitle, Field, FieldLabel, Label, FieldBody, Control, Input, Help } from 'bloomer';

export interface IMonsterStatsProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// parent
	Parent: string,
	// initial values
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterStatsState {
	[key: string]: number | string | undefined;
	// data
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
	// errors
	StrengthError?: string;
	DexterityError?: string;
	ConstitutionError?: string;
	IntelligenceError?: string;
	WisdomError?: string;
	CharismaError?: string;
}

export class MonsterStats extends React.Component<IMonsterStatsProps, IMonsterStatsState> {
	constructor(props: IMonsterStatsProps) {
		super(props);
		this.state = {
			Strength: props.Strength,
			Dexterity: props.Dexterity,
			Constitution: props.Constitution,
			Intelligence: props.Intelligence,
			Wisdom: props.Wisdom,
			Charisma: props.Charisma
		};
	}

	private keyNames = [ "Strength", "Dexterity", "Constitution", "Intelligence", "Wisdom", "Charisma" ];

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleMonsterStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [this.props.Parent, name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
		}

	render() {
		return (
			<Tile className="box" isVertical isParent >
				<Tile isChild render={ (props: any) =>
					<React.Fragment>
						<Subtitle>{this.props.Parent.replace(/([A-Z])/g, ' $1').trim()}</Subtitle>
						{this.keyNames.map((value: string) =>
							<Field isHorizontal key={this.props.Parent+value}>
								<FieldLabel isNormal>
									<Label>{value}</Label>
								</FieldLabel>
								<FieldBody>
									<Field>
										<Control>
											<Input
												disabled={this.props.disabled}
												id={this.props.Parent+value}
												type='number'
												placeholder={value}
												autoComplete={value}
												value={this.state[value] != undefined ? this.state[value] : ''}
												name={value}
												onChange={this.handleMonsterStringChange} />
										</Control>
										<Help isColor='danger' id={this.props.Parent+value}>{this.state[value+'Error']}</Help>
									</Field>
								</FieldBody>
							</Field>
						)}
					</React.Fragment>
				} />
			</Tile>
		);
	}
}