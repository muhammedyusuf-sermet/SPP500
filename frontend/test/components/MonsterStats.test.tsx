import * as React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';

import { IMonsterStatsProps, IMonsterStatsState, MonsterStats } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterStats';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {

	}
};

////// Happy Path //////
describe('Monster Stats', () => {

	let monsterStatsInstance: ReactWrapper<IMonsterStatsProps, IMonsterStatsState, MonsterStats>;

	describe('Provided Stat values', () => {

		beforeEach(() => {
			const payloadSchema = Joi.object({
				AbilityScores: Joi.object({
					Strength: Joi.number().integer().greater(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).label('Charisma')
				})
			});
			monsterStatsInstance =
				mount<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='AbilityScores'
						initial={{
							Strength: 123,
							Dexterity: 53,
							Constitution: 25,
							Intelligence: 85,
							Wisdom: 43,
							Charisma: 32
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterStatsInstance).toBeDefined();
		});

		it('should have the correct number for Strength', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresStrength').props().value).toEqual(123);
		});

		it('should have the correct number for Dexterity', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresDexterity').props().value).toEqual(53);
		});

		it('should have the correct number for Constitution', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresConstitution').props().value).toEqual(25);
		});

		it('should have the correct number for Intelligence', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresIntelligence').props().value).toEqual(85);
		});

		it('should have the correct number for Wisdom', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresWisdom').props().value).toEqual(43);
		});

		it('should have the correct number for Charisma', () => {
			expect(monsterStatsInstance.find('input#AbilityScoresCharisma').props().value).toEqual(32);
		});
	});

	describe('componentWillReceiveProps()', () => {
		beforeEach(() => {
			const payloadSchema = Joi.object({
				AbilityScores: Joi.object({
					Strength: Joi.number().integer().greater(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).label('Charisma')
				})
			});
			monsterStatsInstance =
				mount<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='AbilityScores'
						initial={{
							Strength: 123,
							Dexterity: 53,
							Constitution: 25,
							Intelligence: 85,
							Wisdom: 43,
							Charisma: 32
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterStatsInstance).toBeDefined();
		});

		it('renders new props', () => {
			monsterStatsInstance.setProps({
				initial: {
					Strength: 12,
					Dexterity: 36,
					Constitution: 5,
					Intelligence: 52,
					Wisdom: 71,
					Charisma: 14
				}
			})
			expect(monsterStatsInstance.find('input#AbilityScoresStrength').props().value).toEqual(12);
			expect(monsterStatsInstance.find('input#AbilityScoresDexterity').props().value).toEqual(36);
			expect(monsterStatsInstance.find('input#AbilityScoresConstitution').props().value).toEqual(5);
			expect(monsterStatsInstance.find('input#AbilityScoresIntelligence').props().value).toEqual(52);
			expect(monsterStatsInstance.find('input#AbilityScoresWisdom').props().value).toEqual(71);
			expect(monsterStatsInstance.find('input#AbilityScoresCharisma').props().value).toEqual(14);
		});

		it('renders without crashing with same props', () => {
			monsterStatsInstance.setProps({
				initial: {
					Strength: 123,
					Dexterity: 53,
					Constitution: 25,
					Intelligence: 85,
					Wisdom: 43,
					Charisma: 32
				}
			});
			expect(monsterStatsInstance.find('input#AbilityScoresStrength').props().value).toEqual(123);
			expect(monsterStatsInstance.find('input#AbilityScoresDexterity').props().value).toEqual(53);
			expect(monsterStatsInstance.find('input#AbilityScoresConstitution').props().value).toEqual(25);
			expect(monsterStatsInstance.find('input#AbilityScoresIntelligence').props().value).toEqual(85);
			expect(monsterStatsInstance.find('input#AbilityScoresWisdom').props().value).toEqual(43);
			expect(monsterStatsInstance.find('input#AbilityScoresCharisma').props().value).toEqual(32);
		});
	});

	describe('Small snapshot', () => {
		it ('matches snapshot', () => {
			const payloadSchema = Joi.object({
				SavingThrows: Joi.object({
					Strength: Joi.number().integer().greater(0).allow(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).allow(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).allow(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).allow(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).allow(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).allow(0).label('Charisma')
				})
			});
			const shallowMonsterStatsInstance =
				shallow<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='SavingThrows'
						initial={{}}/>
				);
			expect(shallowMonsterStatsInstance).toMatchSnapshot();
		});
	});

	describe('Provided Stat values', () => {

		beforeEach(() => {
			const payloadSchema = Joi.object({
				SavingThrows: Joi.object({
					Strength: Joi.number().integer().greater(0).allow(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).allow(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).allow(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).allow(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).allow(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).allow(0).label('Charisma')
				})
			});
			monsterStatsInstance =
				mount<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='SavingThrows'
						initial={{
							Strength: 56,
							Dexterity: 0,
							Constitution: 76,
							Intelligence: 13,
							Wisdom: 54,
							Charisma: 10
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterStatsInstance).toBeDefined();
		});

		it('should have the correct number for Strength', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsStrength').props().value).toEqual(56);
		});

		it('should have the correct number for Dexterity', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsDexterity').props().value).toEqual(0);
		});

		it('should have the correct number for Constitution', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsConstitution').props().value).toEqual(76);
		});

		it('should have the correct number for Intelligence', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsIntelligence').props().value).toEqual(13);
		});

		it('should have the correct number for Wisdom', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsWisdom').props().value).toEqual(54);
		});

		it('should have the correct number for Charisma', () => {
			expect(monsterStatsInstance.find('input#SavingThrowsCharisma').props().value).toEqual(10);
		});
	});

	describe('Happy Path Ability Score', () => {

		beforeEach(() => {
			const payloadSchema = Joi.object({
				AbilityScores: Joi.object({
					Strength: Joi.number().integer().greater(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).label('Charisma')
				})
			});
			monsterStatsInstance =
				mount<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='AbilityScores'
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterStatsInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterStatsInstance.find('input#AbilityScoresStrength').simulate('change', { target: { value: 17 } });
			monsterStatsInstance.find('input#AbilityScoresDexterity').simulate('change', { target: { value: 15 } });
			monsterStatsInstance.find('input#AbilityScoresConstitution').simulate('change', { target: { value: 13 } });
			monsterStatsInstance.find('input#AbilityScoresIntelligence').simulate('change', { target: { value: 12 } });
			monsterStatsInstance.find('input#AbilityScoresWisdom').simulate('change', { target: { value: 16 } });
			monsterStatsInstance.find('input#AbilityScoresCharisma').simulate('change', { target: { value: 15 } });
			expect(monsterStatsInstance.state()).toEqual({
				// data
				Strength: 17,
				Dexterity: 15,
				Constitution: 13,
				Intelligence: 12,
				Wisdom: 16,
				Charisma: 15,
				// errors
				StrengthError: undefined,
				DexterityError: undefined,
				ConstitutionError: undefined,
				IntelligenceError: undefined,
				WisdomError: undefined,
				CharismaError: undefined,
			});
		});
		// STATE CHANGE
		it('should change only Strength when Strength is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresStrength').simulate('change', { target: { value: '22' } })
			expect(monsterStatsInstance.state().Strength).toEqual(22);
		})

		it('should change only Dexterity when Dexterity is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresDexterity').simulate('change', { target: { value: '74' } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(74);
		})

		it('should change only Constitution when Constitution is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresConstitution').simulate('change', { target: { value: '5' } })
			expect(monsterStatsInstance.state().Constitution).toEqual(5);
		})

		it('should change only Intelligence when Intelligence is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresIntelligence').simulate('change', { target: { value: '91' } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(91);
		})

		it('should change only Wisdom when Wisdom is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresWisdom').simulate('change', { target: { value: '24' } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(24);
		})

		it('should change only Charisma when Charisma is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresCharisma').simulate('change', { target: { value: '100' } })
			expect(monsterStatsInstance.state().Charisma).toEqual(100);
		})

		// DEFAULTS
		it('should change Strength back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresStrength').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Strength).toEqual(undefined);
		})

		it('should change Dexterity back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresDexterity').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(undefined);
		})

		it('should change Constitution back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresConstitution').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Constitution).toEqual(undefined);
		})

		it('should change Intelligence back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresIntelligence').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(undefined);
		})

		it('should change Wisdom back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresWisdom').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(undefined);
		})

		it('should change Charisma back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#AbilityScoresCharisma').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Charisma).toEqual(undefined);
		})

		// VALIDATION
		it('should validate Strength when Strength is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresStrength').simulate('change', { target: { value: '-12'  } })
			expect(monsterStatsInstance.state().Strength).toEqual(-12);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresStrength-helper-text').text()).toEqual('\"Strength\" must be greater than 0');
		})

		it('should validate Dexterity when Dexterity is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresDexterity').simulate('change', { target: { value: '-16'  } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(-16);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresDexterity-helper-text').text()).toEqual('\"Dexterity\" must be greater than 0');
		})

		it('should validate Constitution when Constitution is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresConstitution').simulate('change', { target: { value: '-6'  } })
			expect(monsterStatsInstance.state().Constitution).toEqual(-6);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresConstitution-helper-text').text()).toEqual('\"Constitution\" must be greater than 0');
		})

		it('should validate Intelligence when Intelligence is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresIntelligence').simulate('change', { target: { value: '-102'  } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(-102);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresIntelligence-helper-text').text()).toEqual('\"Intelligence\" must be greater than 0');
		})

		it('should validate Wisdom when Wisdom is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresWisdom').simulate('change', { target: { value: '0'  } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(0);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresWisdom-helper-text').text()).toEqual('\"Wisdom\" must be greater than 0');
		})

		it('should validate Charisma when Charisma is changed', () => {
			monsterStatsInstance.find('input#AbilityScoresCharisma').simulate('change', { target: { value: '-9'  } })
			expect(monsterStatsInstance.state().Charisma).toEqual(-9);
			expect(monsterStatsInstance.find('FormHelperText#AbilityScoresCharisma-helper-text').text()).toEqual('\"Charisma\" must be greater than 0');
		})
	});

	describe('Happy Path Saving Throws', () => {

		beforeEach(() => {
			const payloadSchema = Joi.object({
				SavingThrows: Joi.object({
					Strength: Joi.number().integer().greater(0).allow(0).label('Strength'),
					Dexterity: Joi.number().integer().greater(0).allow(0).label('Dexterity'),
					Constitution: Joi.number().integer().greater(0).allow(0).label('Constitution'),
					Intelligence: Joi.number().integer().greater(0).allow(0).label('Intelligence'),
					Wisdom: Joi.number().integer().greater(0).allow(0).label('Wisdom'),
					Charisma: Joi.number().integer().greater(0).allow(0).label('Charisma')
				})
			});
			monsterStatsInstance =
				mount<MonsterStats, IMonsterStatsProps, IMonsterStatsState>(
					<MonsterStats
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						Parent='SavingThrows'
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterStatsInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterStatsInstance.find('input#SavingThrowsStrength').simulate('change', { target: { value: 23 } });
			monsterStatsInstance.find('input#SavingThrowsDexterity').simulate('change', { target: { value: 0 } });
			monsterStatsInstance.find('input#SavingThrowsConstitution').simulate('change', { target: { value: 1 } });
			monsterStatsInstance.find('input#SavingThrowsIntelligence').simulate('change', { target: { value: 100 } });
			monsterStatsInstance.find('input#SavingThrowsWisdom').simulate('change', { target: { value: 17 } });
			monsterStatsInstance.find('input#SavingThrowsCharisma').simulate('change', { target: { value: 0 } });
			expect(monsterStatsInstance.state()).toEqual({
				// data
				Strength: 23,
				Dexterity: 0,
				Constitution: 1,
				Intelligence: 100,
				Wisdom: 17,
				Charisma: 0,
				// errors
				StrengthError: undefined,
				DexterityError: undefined,
				ConstitutionError: undefined,
				IntelligenceError: undefined,
				WisdomError: undefined,
				CharismaError: undefined,
			});
		});
		// STATE CHANGE
		it('should change only Strength when Strength is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsStrength').simulate('change', { target: { value: '22' } })
			expect(monsterStatsInstance.state().Strength).toEqual(22);
		})

		it('should change only Dexterity when Dexterity is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsDexterity').simulate('change', { target: { value: '0' } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(0);
		})

		it('should change only Constitution when Constitution is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsConstitution').simulate('change', { target: { value: '5' } })
			expect(monsterStatsInstance.state().Constitution).toEqual(5);
		})

		it('should change only Intelligence when Intelligence is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsIntelligence').simulate('change', { target: { value: '0' } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(0);
		})

		it('should change only Wisdom when Wisdom is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsWisdom').simulate('change', { target: { value: '24' } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(24);
		})

		it('should change only Charisma when Charisma is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsCharisma').simulate('change', { target: { value: '100' } })
			expect(monsterStatsInstance.state().Charisma).toEqual(100);
		})

		// DEFAULTS
		it('should change Strength back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsStrength').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Strength).toEqual(undefined);
		})

		it('should change Dexterity back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsDexterity').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(undefined);
		})

		it('should change Constitution back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsConstitution').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Constitution).toEqual(undefined);
		})

		it('should change Intelligence back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsIntelligence').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(undefined);
		})

		it('should change Wisdom back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsWisdom').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(undefined);
		})

		it('should change Charisma back to undefined when input is empty', () => {
			monsterStatsInstance.find('input#SavingThrowsCharisma').simulate('change', { target: { value: '' } })
			expect(monsterStatsInstance.state().Charisma).toEqual(undefined);
		})

		// VALIDATION
		it('should validate Strength when Strength is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsStrength').simulate('change', { target: { value: '-12'  } })
			expect(monsterStatsInstance.state().Strength).toEqual(-12);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsStrength-helper-text').text()).toEqual('\"Strength\" must be greater than 0');
		})

		it('should validate Dexterity when Dexterity is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsDexterity').simulate('change', { target: { value: '-16'  } })
			expect(monsterStatsInstance.state().Dexterity).toEqual(-16);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsDexterity-helper-text').text()).toEqual('\"Dexterity\" must be greater than 0');
		})

		it('should validate Constitution when Constitution is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsConstitution').simulate('change', { target: { value: '-6'  } })
			expect(monsterStatsInstance.state().Constitution).toEqual(-6);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsConstitution-helper-text').text()).toEqual('\"Constitution\" must be greater than 0');
		})

		it('should validate Intelligence when Intelligence is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsIntelligence').simulate('change', { target: { value: '-102'  } })
			expect(monsterStatsInstance.state().Intelligence).toEqual(-102);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsIntelligence-helper-text').text()).toEqual('\"Intelligence\" must be greater than 0');
		})

		it('should validate Wisdom when Wisdom is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsWisdom').simulate('change', { target: { value: '-1'  } })
			expect(monsterStatsInstance.state().Wisdom).toEqual(-1);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsWisdom-helper-text').text()).toEqual('\"Wisdom\" must be greater than 0');
		})

		it('should validate Charisma when Charisma is changed', () => {
			monsterStatsInstance.find('input#SavingThrowsCharisma').simulate('change', { target: { value: '-9'  } })
			expect(monsterStatsInstance.state().Charisma).toEqual(-9);
			expect(monsterStatsInstance.find('FormHelperText#SavingThrowsCharisma-helper-text').text()).toEqual('\"Charisma\" must be greater than 0');
		})
	});
});