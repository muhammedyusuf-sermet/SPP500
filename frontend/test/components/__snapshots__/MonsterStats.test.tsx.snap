// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster Stats Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Saving Throws
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsStrength"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsStrength"
          >
            Strength
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Strength-helper-text"
            id="SavingThrowsStrength"
            name="Strength"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsStrength-helper-text"
          />
          <withHelpersModifiers(Help)>
            Strength
            : 
            measuring physical power
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsDexterity"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsDexterity"
          >
            Dexterity
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Dexterity-helper-text"
            id="SavingThrowsDexterity"
            name="Dexterity"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsDexterity-helper-text"
          />
          <withHelpersModifiers(Help)>
            Dexterity
            : 
            measuring agility
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsConstitution"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsConstitution"
          >
            Constitution
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Constitution-helper-text"
            id="SavingThrowsConstitution"
            name="Constitution"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsConstitution-helper-text"
          />
          <withHelpersModifiers(Help)>
            Constitution
            : 
            measuring endurance
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsIntelligence"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsIntelligence"
          >
            Intelligence
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Intelligence-helper-text"
            id="SavingThrowsIntelligence"
            name="Intelligence"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsIntelligence-helper-text"
          />
          <withHelpersModifiers(Help)>
            Intelligence
            : 
            measuring reasoning and memory
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsWisdom"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsWisdom"
          >
            Wisdom
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Wisdom-helper-text"
            id="SavingThrowsWisdom"
            name="Wisdom"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsWisdom-helper-text"
          />
          <withHelpersModifiers(Help)>
            Wisdom
            : 
            measuring Perception and Insight
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SavingThrowsCharisma"
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SavingThrowsCharisma"
          >
            Charisma
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Charisma-helper-text"
            id="SavingThrowsCharisma"
            name="Charisma"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SavingThrowsCharisma-helper-text"
          />
          <withHelpersModifiers(Help)>
            Charisma
            : 
            measuring force of personality
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
